### ISSUE TITLE
rclone backup skips Windows junctions silently

### ISSUE THREAD

**Issue by cowwoc** [restic/restic#3674]
Output of `restic version`
--------------------------
restic 0.12.1 compiled with go1.16.6 on windows/amd64

How did you run restic exactly?
-------------------------------

restic backup --use-fs-snapshot -r b2:my-repo --files-from="include.txt" --iexclude-file="exclude.txt" 

where `include.txt` contains:

```
c:/test
```

and `exclude.txt` is empty, and c:\test contains windows junctions created using `mklink /d source target`.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

b2

Expected behavior
-----------------

restic should back up all directories recursively

Actual behavior
---------------

restic backs up all normal sub-directories and silently skips over any windows junctions.

Steps to reproduce the behavior
-------------------------------

Back up any directory containing windows junctions created using `mklink /d source target`.

Do you have any idea what may have caused this?
-----------------------------------------------

No

Do you have an idea how to solve the issue?
-------------------------------------------

No

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

**Comment by greatroar** [comment]
The title says rclone backup, but you clearly mean restic backup.

**Comment by cowwoc** [comment]
@greatroar Good catch :) Fixed.

**Comment by MichaelEischer** [comment]
According to the documentation `mklink /d` creates a symbolic link to a directory. restic _always_ backs up symbolic links as symbolic links and _never_ follows them.

What's a bit surprising is that directory junctions created using `mklink /j` are also considered to be symlinks.

**Comment by MichaelEischer** [comment]
So essentially the question is whether directory junctions should be treated like mount points on Unix systems or not. At least Go has  decided that both symlink and junctions should be treated as symlinks. And judging from the fact that the junctions seem to behave somewhat similar to symlinks that appears to be a reasonable decision.

**Comment by Dobatymo** [comment]
> And judging from the fact that the junctions seem to behave somewhat similar to symlinks that appears to be a reasonable decision.
The main difference is on network file systems. Junctions are resolved on the serverside, while symlink are resolved on the clientside.

**Comment by cowwoc** [comment]
> restic always backs up symbolic links as symbolic links and never follows them.

Maybe I'm wrong, but I have a feeling that this behavior should be configurable. At least in my case, I'd like all symlinks to get backed up as symlinks but also for restic to follow them.

**Comment by MichaelEischer** [comment]
Looks like I forgot to link to the existing discussions around symlinks. What you're asking for is essential part 2 of https://github.com/restic/restic/issues/542#issuecomment-328263959 . That is include the symlinks _and_ their targets.

> Junctions are resolved on the serverside, while symlink are resolved on the clientside.

When Junctions are resolved on the server-side are these then transparent to a client?

**Comment by Dobatymo** [comment]
> When Junctions are resolved on the server-side are these then transparent to a client?

Yes. See for example <https://superuser.com/a/1479723>.

**Comment by MichaelEischer** [comment]
We'll if the junction is followed transparently that, begs the question whether restic would even notice that the junction exists or not. If a junction via Samba just looks like a normal directory, then it would already be backed up and then we could just treat junctions and symlinks the same.

**Comment by cowwoc** [comment]
It seems my original bug report used the wrong terminology. I am using `mklink /d` to create symlinks under Windows and restic does not follow it. If Junctions already work, can I just rename this issue to make it about recursing into symlinks?

**Comment by Dobatymo** [comment]
@MichaelEischer yes if you do backup from a samba share that's true. But if restic is run on the share server itself the behaviour would be different. I think that is confusing.

**Comment by MichaelEischer** [comment]
This might be fixed by Go 1.23. From the [changelog](https://go.dev/doc/go1.23#minor_library_changes):

> On Windows, the mode bits reported by [Lstat](https://go.dev/pkg/os#Lstat) and [Stat](https://go.dev/pkg/os#Stat) for reparse points changed. Mount points no longer have [ModeSymlink](https://go.dev/pkg/os#ModeSymlink) set, and reparse points that are not symlinks, Unix sockets, or dedup files now always have [ModeIrregular](https://go.dev/pkg/os#ModeIrregular) set. This behavior is controlled by the winsymlink setting. For Go 1.23, it defaults to winsymlink=1. Previous versions default to winsymlink=0.
